### PR TITLE
control/controlbase: use less memory when idle (remove rxState.buf)

### LINKED ISSUE #4113: control/controlbase: reduce memory usage of `Conn`
The `controlbase.Conn` is currently:

```go
const (
        // maxMessageSize is the maximum size of a protocol frame on the
        // wire, including header and payload.
        maxMessageSize = 4096
)

type Conn struct {
        conn          net.Conn
        version       uint16
        peer          key.MachinePublic
        handshakeHash [blake2s.Size]byte
        rx            rxState
        tx            txState
}

// rxState is all the Conn state that Read uses.
type rxState struct {
        sync.Mutex
        cipher    cipher.AEAD
        nonce     nonce
        buf       [maxMessageSize]byte
        n         int    // number of valid bytes in buf
        next      int    // offset of next undecrypted packet
        plaintext []byte // slice into buf of decrypted bytes
}

// txState is all the Conn state that Write uses.
type txState struct {
        sync.Mutex
        cipher cipher.AEAD
        nonce  nonce
        buf    [maxMessageSize]byte
        err    error // records the first partial write error for all future calls
}
```

This is a tracking bug to investigate reducing the size of Conn, which is currently over 8KB.

This doesn't matter for the client, but it'll matter server-side with tons of connections.

I suspect most most of the tx pools are empty at any moment, for instance. That halves the size.

And for reading, we probably don't need the recv buf either until the 3 byte header read unblocks.

So we can pool both.

For #3488

/cc @danderson @maisem @josharian @crawshaw

### PR DESCRIPTION
**Pull request by bradfitz**
Uses 4KB less per Conn.

Fixes #4113

### CODE REVIEW

**Comment by bradfitz** [comment]
I had to mail this quickly so haven't given it my usual once-over. Will do later before submitting, but just in case you want to review it early....

File: control/controlbase/conn.go [hunk 1]
```diff
@@ -96,8 +97,26 @@ func (c *Conn) readNLocked(total int) ([]byte, error) {
 		if total <= c.rx.n {
 			return c.rx.buf[:total], nil
 		}
-
-		n, err := c.conn.Read(c.rx.buf[c.rx.n:])
+		var n int
+		var err error
+		if c.rx.buf == nil && c.rx.n == 0 && total == headerLen {
```
Comment (danderson) [inline control/controlbase/conn.go:102]: This branch is suspect: if one of the latter two conditions is false, the else would end up reading into a nil buffer. So either the latter conditions aren't necessary and are always true here, or there are more alternatives needed?

**Review by danderson — COMMENTED** [review]
Brain's mush today, so only did a very quick pass. I need to re-read more carefully tomorrow, the buffer dance is much trickier in the read path.

File: control/controlbase/conn.go (hunk 1 above)
Comment (bradfitz) [reply control/controlbase/conn.go]: Tightened it up. PTAL.

File: control/controlbase/conn.go [hunk 2]
```diff
@@ -96,8 +101,29 @@ func (c *Conn) readNLocked(total int) ([]byte, error) {
...
+			}
+			// Optimization to reduce memory usage.
+			// Most connections are blocked forever waiting for
+			// a read, so we don't want c.rx.buf to be allocated until
+			// we know there's data to read. Instead, when we're
+			// waiting for data to arrive here, read into the
+			// 3 byte hdrBuf:
+			n, err = c.conn.Read(c.rx.hdrBuf[:])
+			if n > 0 {
+				if n > headerLen {
```
Comment (danderson) [inline control/controlbase/conn.go:118]: This check's probably over-defensive, if net.Conn.Read ever returns n > buf, we have big problems :)

File: control/controlbase/conn.go [hunk 3]
```diff
@@ -190,6 +216,14 @@ func (c *Conn) decryptOneLocked() error {
 		c.rx.next = 0
 	}
 
+	// Return out buffer to the pool if it's empty, lest we be
```
Comment (danderson) [inline control/controlbase/conn.go:219]: typo: out > our? I think? Unless this was copy pasted from the tx path and you meant "the out buffer", but this isn't an out buffer.

File: control/controlbase/conn_test.go [hunk 4]
```diff
@@ -224,6 +226,55 @@ func TestConnStd(t *testing.T) {
...
+	if testing.Short() {
+		num = 100
+	}
+	runtime.GC()
+	var ms0 runtime.MemStats
+	runtime.ReadMemStats(&ms0)
+
+	for i := 0; i < num; i++ {
+		client, server := pair(t)
+		defer server.Close()
```
Comment (danderson) [inline control/controlbase/conn_test.go:245]: no client.Close?

**Review by danderson — APPROVED** [review]

File: control/controlbase/conn.go (hunk 2 above)
Comment (bradfitz) [reply control/controlbase/conn.go:118]: Removed. It'll panic two lines later if so anyway.

File: control/controlbase/conn_test.go (hunk 4 above)
Comment (bradfitz) [reply control/controlbase/conn_test.go]: done

**Pull request merged**